### DISCUSSION TITLE
"Create Modular / Multi Actions"

### DISCUSSION THREAD

**Discussion by Data-007** [streamdeck-linux-gui/streamdeck-linux-gui#42] [Ideas / Feature Requests]
I would like to see modular actions done as previously mulled over by Pieter(doddgy rabbit)
I think this is even mentioned in one of  the threads in issues(for clarity it should be proposed here I think)

The general idea was discussed in the previous GitHub repo by Pieter on/in Timothy Crosleys GitHub.

 Create modular actions #208 

I would like to see that suggestion here on this project it was laid out in grate detail. A picture speaks a thousand words.
![ModAction](https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/assets/53057579/6d271fa5-5268-481e-89b2-2fa1dbe8112b)
This image was taken from the previous GitHub its not mine.

So much I see in this that a number of the ideas proposed - discussed in this GitHub could all benefit from it.

I apologize if this is a duplicate I thought someone mentioned it on here somewhere I cant find it.

**Comment by Danie10** [comment]
I'm wondering if it was not described somewhere as multi-actions?

**Comment by zutmkr** [comment]
Hello, 
This as @Danie10 wrote look like a multi-actions to me as well. By "Plugins" I would understand that we have one the right side of the UI something like this.

![image](https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/assets/15087930/ef93b34b-db50-429a-9906-6bec451a506e)

Similar to org SW, so you can load an new module and expand its functions.

Also I'm new here, so hello to everyone :)

**Reply by Data-007** [reply]
Keep it coming if you look at the link I posted in my original post its(the layout/idea) is discussed in grate detail. 

@Danie10 Is this a repeat of a post in here? I seem to remember two attempts issue's at the old GitHub, I seem to mind 
something here to I think. But I cant locate it Ive re-scanned visually the subjects here.
#sftim

**Reply by Danie10** [reply]
@Data-007 no I don't see it logged in this project as new features. I think it may have been on the old project. I'm not sure why it would need to be a plugin though, and I'd rather leave that up to the devs whether it can go into main code, or needs a separate plugin. We should probably just suggest / request any functionality we'd like to use.  Can maybe make the subject line:

"Create Modular / Multi Actions"

**Reply by Data-007** [reply]
@Danie10 rename how? im not as proficient as some, ive gone and tried editing the first post that did not work.

Plugin
What I had in mind was a feature that allowed a wider range of actions, at the moment one can, launch a program, output text, press keys. Three actions I would propose that the number off actions be increased and be able to be laid out in a script like format(do this then that then the next). This was explained in the old repo so much better as I can explain here.(see first-post for link)
I guess its the functions that one could do, that could be linked together one after another without having to have the hassle of programing in python or shell script. if this then that..

Take your (ffff) idea function repeated while button X is pushed, this could be another action that you could do in order along with other items when the key is no longer pressed. This is putting it in basic form's, making it as easy as selecting a number of item in steps to be done procedural like a basic script language.  Each script could be considered a plugin.

Combine this with the ability of being able to have "dynamic display on buttons" #5  mainly and other bits suggested on this repo.

**Comment by Danie10** [comment]
@Data-007 OK I see you found the rename / edit option.

As far as scripting goes, that can already be done with the command option. I created a small bash script to do something, and the button just executes that script (temporarily at least for now).

I never had a good look at what the Windows version was doing with multi-action, but guessed it was to do with firing off maybe two or three commands, etc. I know we can already string key presses together.

**Reply by Data-007** [reply]
@zutmkr I like or liked the term plugin's for a reason. 
At the moment to my knowledge we can do a number of things already. we cant determine the **sequence** of events(unless in a script)  this is why I chose the term 'plugin'. I like your graphic.

@Danie10 chose to request a name change to 'Multi-actions' I am not sure either how the windows one works(lets not go to windows).  **Multi**-actions this I can understand, we cant for instance unless in a script determine **sequence**. The original
interface we have now is great, simplicity, but an 'all or nothing' no sequence.

**modular** little pockets of actions. I get the script idea, I guess the term fits to a degree, I come back to the word sequence.

_Tried to keep this on-track and constructive_

**Reply by ClassickTech** [reply]
How are you doing this? Is there anyway to slow down the text? It's typing to fast so it's cutting off the output. That's why I wish there was a delay setting for Commands and Wrrite Text

**Reply by Danie10** [reply]
@ClassickTech if you incorporate something like xdotool as a command, it has the ability to insert delays between commands e.g. `xdotool type --delay 100 hello`. Not sure if that helps.
